fix(types): add ResultSet interface and improve isResultSet typing

ranidam opened the pull request:
### This PR introduces proper TypeScript support for the `ResultSet` returned by `evaluate()` when executing multiple expressions in MathJS.

### Changes
1) Added a `ResultSet` interface to `types/index.d.ts`
2) Updated the `isResultSet` type guard to use a type predicate `(x is ResultSet)`
3) Included a type-level test in `test/typescript-tests/testTypes.ts` to validate inference


When using `evaluate()` with multiple expressions (e.g.,`1 + 1; 2 + 2;`),   MathJS returns a `ResultSet.` 
However:
a) The `ResultSet` type was not previously exported
b) `isResultSet()` did not indicate that the input is a `ResultSet`
c) This caused issues when trying to type and handle `evaluate()` results safely

This PR fixes those gaps, allowing TypeScript developers to detect and type `ResultSet` results reliably.

Let me know if you'd like me to adjust anything further!

josdejong commented on types/index.d.ts:
```diff
@@ -4319,6 +4321,17 @@ export interface EvalFunction {
+  entries: unknown[];
+  valueOf(): unknown[];
+  toString(): string;
+  [key: string]: any;
```
I think the line `[key: string]: any;` shouldn't be there.

ranidam commented on types/index.d.ts:
```diff
@@ -4319,6 +4321,17 @@ export interface EvalFunction {
+export interface ResultSet {
+  entries: unknown[];
+  valueOf(): unknown[];
+  toString(): string;
```
Absolutely! I’ll add the `toJSON(): unknown;` method to the `ResultSet` interface and push the update along with the other changes.

ranidam commented on test/typescript-tests/testTypes.ts:
```diff
@@ -172,6 +172,16 @@ Bignumbers examples
   }
 }
 
+import { evaluate, isResultSet } from 'mathjs';
```
I’ll move the `evaluate` and `isResultSet` imports to the existing `mathjs` import at the top of the file for consistency. Will include that in the next commit.

ranidam commented:
>  I've applied all requested changes: removed the `[key: string]: any;`, added `toJSON()`, cleaned up the unused variable, and ran  `npm run format`. Moved the `evaluate` and `isResultSet` imports to the existing `mathjs`. Let me know if anything else is needed."

josdejong commented:
Thanks Rani, all good to go now 👌

josdejong commented:
Your fix is published now in `v14.5.2`, thanks again!
